source/extensions/stat_sinks/hystrix/hystrix.cc

```diff
@@ -0,0 +1,375 @@
+#include "extensions/stat_sinks/hystrix/hystrix.h"
+
+#include <chrono>
+#include <ctime>
+#include <iostream>
+#include <sstream>
+
+#include "common/buffer/buffer_impl.h"
+#include "common/common/logger.h"
+#include "common/http/headers.h"
+
+#include "absl/strings/str_cat.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace StatSinks {
+namespace Hystrix {
+
+const uint64_t HystrixSink::DEFAULT_NUM_OF_BUCKETS;
+
+ClusterStatsCache::ClusterStatsCache(const std::string& cluster_name) {
+  const std::string cluster_name_with_prefix = absl::StrCat("cluster.", cluster_name, ".");
+
+  upstream_rq_2xx_name_ = absl::StrCat(cluster_name_with_prefix, "upstream_rq_2xx");
+  upstream_rq_4xx_name_ = absl::StrCat(cluster_name_with_prefix, "upstream_rq_4xx");
+  retry_upstream_rq_4xx_name_ = absl::StrCat(cluster_name_with_prefix, "retry.upstream_rq_4xx");
+  upstream_rq_5xx_name_ = absl::StrCat(cluster_name_with_prefix, "upstream_rq_5xx");
+  retry_upstream_rq_5xx_name_ = absl::StrCat(cluster_name_with_prefix, "retry.upstream_rq_5xx");
+
+  errors_name_ = absl::StrCat(cluster_name_with_prefix, "errors");
+  success_name_ = absl::StrCat(cluster_name_with_prefix, "success");
+  total_name_ = absl::StrCat(cluster_name_with_prefix, "total");
+  timeouts_name_ = absl::StrCat(cluster_name_with_prefix, "timeouts");
+  rejected_name_ = absl::StrCat(cluster_name_with_prefix, "rejected");
+}
+
+// Add new value to rolling window, in place of oldest one.
+void HystrixSink::pushNewValue(RollingWindow& rolling_window, uint64_t value) {
+  if (rolling_window.empty()) {
+    rolling_window.resize(window_size_, value);
+  } else {
+    rolling_window[current_index_] = value;
+  }
+}
+
+uint64_t HystrixSink::getRollingValue(RollingWindow rolling_window) {
+
+  if (rolling_window.empty()) {
+    return 0;
+  }
+  // If the counter was reset, the result is negative
+  // better return 0, will be back to normal once one rolling window passes.
+  if (rolling_window[current_index_] < rolling_window[(current_index_ + 1) % window_size_]) {
+    return 0;
+  } else {
+    return rolling_window[current_index_] - rolling_window[(current_index_ + 1) % window_size_];
+  }
+}
+
+void HystrixSink::updateRollingWindowMap(Upstream::ClusterInfoConstSharedPtr cluster_info,
+                                         Stats::Store& stats) {
```

Comment:
We should be using each cluster's stats scope to get cluster-specific stats.  Otherwise, they will not be deallocated when the cluster is removed.

Reply:
Not sure I understand the comment about not being deallocated when the cluster is removed. Are you referring to using `Upstream::ClusterStats`? or the Shared pointer `Upstream::ClusterInfoConstSharedPtr`?

Reply:
We should be using ClusterInfo::statsScope() - see [here](https://github.com/envoyproxy/envoy/blob/54065c13209f774e5f425293a883b2f1b404cded/include/envoy/upstream/upstream.h#L501). 

The stat scope system is a little confusing with respect to how it manages memory.  When you request a stat from a scope using `scope.counters(stat_name)`, the scope will then ensure the underlying stat stays alive as long as that scope is alive.  If we pull the cluster stats from the server scope, that means the stats that we're requesting are required to live as long as the server's stats scope, which is the life of the Envoy instance.  If we grab them from the cluster scope, they will be deallocated when the cluster is removed as one would expect for cluster-specific stats.

Reply:
Is it new interface? It is very good for us, since using this, we won't need to save the concatenated string of the statistic's name per cluster. What about the stats we are currently getting from `Upstream::ClusterStats::stats()`? Take them also from the `Upstream::ClusterInfo::statsScope()` instead?

Reply:
Yup, it should make things much easier.  Technically, you could pull any stats in `Upstream::ClusterStats::stats()` from `Upstream::ClusterInfo::statsScope()`, but it's better the way you have it. `Upstream::ClusterStats::stats()` is essentially a struct holding stats from `Upstream::ClusterInfo::statsScope()` to access them without doing a map lookup.